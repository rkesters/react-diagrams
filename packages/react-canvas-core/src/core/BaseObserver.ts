import { has, isNil } from 'lodash';
import { Toolkit } from '../Toolkit';

import debug from 'debug';

const dbg = debug('reactDiagrams:canvasCore:BaseObserver');
export interface BaseEvent {
	firing: boolean;
	stopPropagation: () => any;
}

export interface BaseEventProxy extends BaseEvent {
	function: string;
}

/**
 * Listeners are always in the form of an object that contains methods that take events
 */
export type BaseListener = {
	/**
	 * Generic event that fires before a specific event was fired
	 */
	eventWillFire?: (event: BaseEvent & { function: string }) => void;

	/**
	 * Generic event that fires after a specific event was fired (even if it was consumed)
	 */
	eventDidFire?: (event: BaseEvent & { function: string }) => void;
	/**
	 * Type for other events that will fire
	 */
	[key: string]: (event: BaseEvent) => any;
};

export interface ListenerHandle {
	/**
	 * Used to degister the listener
	 */
	deregister: () => any;
	/**
	 * Original ID of the listener
	 */
	id: string;

	/**
	 * Original Listener
	 */
	listener: BaseListener;
}

export function isaListenerHandle(value: unknown): value is ListenerHandle {
	return has(value, 'id') && has(value, 'deregister') && has(value, 'listener');
}

/**
 * Base observer pattern class for working with listeners
 */
export class BaseObserver<L extends BaseListener = BaseListener> {
	protected listeners: { [id: string]: L };

	constructor() {
		this.listeners = {};
	}

	private fireEventInternal(fire: boolean, k: keyof L, event: BaseEvent) {
		dbg(`fireEventInternal ${k} ${fire}`);
		this.iterateListeners((listener) => {
			dbg(`iterateListeners ${k} ${fire}`);
			// returning false here will instruct itteration to stop
			if (!fire && !event.firing) {
				dbg(`iterateListeners ${k} will not fire ${fire} ${event.firing}`);

				return false;
			}

			dbg(`iterateListeners ${k} ${isNil(listener[k])}`);
			// fire selected listener
			if (listener[k]) {
				listener[k](event as BaseEvent);
			}
		});
	}

	fireEvent<K extends keyof L>(event: Partial<Parameters<L[K]>[0]>, k: keyof L) {
		dbg(`fireEvent ${k}`);
		event = {
			firing: true,
			stopPropagation: () => {
				event.firing = false;
			},
			...event
		};

		// fire pre
		this.fireEventInternal(true, 'eventWillFire', {
			...event,
			function: k
		} as BaseEventProxy);

		dbg(`fireEvent ${k} Main`);
		// fire main event
		this.fireEventInternal(false, k, event as BaseEvent);
		dbg(`fireEvent ${k} post`);
		// fire post
		this.fireEventInternal(true, 'eventDidFire', {
			...event,
			function: k
		} as BaseEventProxy);
	}

	iterateListeners(cb: (listener: L) => any) {
		dbg(`iterateListeners (${Object.keys(this.listeners ?? {}).join(',')})`);
		for (let id in this.listeners) {
			const res = cb(this.listeners[id]);
			// cancel itteration on false
			if (res === false) {
				return;
			}
		}
	}

	getListenerHandle(listener: L): ListenerHandle {
		for (let id in this.listeners) {
			if (this.listeners[id] === listener) {
				return {
					id: id,
					listener: listener,
					deregister: () => {
						delete this.listeners[id];
					}
				};
			}
		}
	}

	registerListener(listener: L): ListenerHandle {
		const id = Toolkit.UID();
		this.listeners[id] = listener;
		const modelID = (this as any).getID ? (this as any).getID() : 'no id';
		dbg(`registerListener on ${modelID} : (${Object.keys(this.listeners ?? {}).join(',')}) (${Object.keys(listener ?? {}).join(',')})`);
		return {
			id: id,
			listener: listener,
			deregister: () => {
			dbg(`deregister on ${modelID} (${Object.keys(this.listeners ?? {}).join(',')}) (${Object.keys(listener ?? {}).join(',')})`);
				delete this.listeners[id];
			}
		};
	}

	deregisterListener(listener: L | ListenerHandle) {
		if (isaListenerHandle(listener)) {
			(listener as ListenerHandle).deregister();
			return true;
		}
		const handle = this.getListenerHandle(listener);
		if (handle) {
			dbg(`deregisterListener (${handle.id})`);

			handle.deregister();
			return true;
		}
		return false;
	}
}
