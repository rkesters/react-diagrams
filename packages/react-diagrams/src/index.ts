import {
	DefaultDiagramState,
	DiagramEngine,
	DiagramModel,
	LinkLayerFactory,
	NodeLayerFactory
} from '@projectstorm/react-diagrams-core';
import {
	DefaultLabelFactory,
	DefaultLinkFactory,
	DefaultNodeFactory,
	DefaultPortFactory
} from '@projectstorm/react-diagrams-defaults';
import { PathFindingLinkFactory } from '@projectstorm/react-diagrams-routing';
import { SelectionBoxLayerFactory, CanvasEngineOptions } from '@projectstorm/react-canvas-core';

export * from '@projectstorm/react-diagrams-core';
export * from '@projectstorm/react-diagrams-defaults';
export * from '@projectstorm/react-diagrams-routing';

/**
 * Construct an engine with the defaults installed
 */
export default (model: DiagramModel, options: CanvasEngineOptions = {}): DiagramEngine => {
	const engine = new DiagramEngine(model,options);

	// register model factories
	engine.getLayerFactories().registerFactory(new NodeLayerFactory(engine));
	engine.getLayerFactories().registerFactory(new LinkLayerFactory(engine));
	engine.getLayerFactories().registerFactory(new SelectionBoxLayerFactory(engine));

	engine.getLabelFactories().registerFactory(new DefaultLabelFactory(engine));
	engine.getNodeFactories().registerFactory(new DefaultNodeFactory(engine)); // i cant figure out why
	engine.getLinkFactories().registerFactory(new DefaultLinkFactory(engine));
	engine.getLinkFactories().registerFactory(new PathFindingLinkFactory(engine));
	engine.getPortFactories().registerFactory(new DefaultPortFactory(engine));

	// register the default interaction behaviours
	engine.getStateMachine().pushState(new DefaultDiagramState(engine));
	return engine;
};
