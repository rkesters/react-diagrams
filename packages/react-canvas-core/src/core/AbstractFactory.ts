import { CanvasEngine } from '../CanvasEngine';
import { FactoryBank } from './FactoryBank';

/**
 * Base factory for all the different types of entities.
 * Gets registered with the engine, and is used to generate models
 */
export abstract class AbstractFactory<E extends CanvasEngine = CanvasEngine> {

	protected bank: FactoryBank | undefined | null;

	constructor(protected type: string, protected engine: E) {
		this.type = type;
	}

	setDiagramEngine(engine: E): void {
		this.engine = engine;
	}

	setFactoryBank(bank: FactoryBank<any> | null): void {
		this.bank = bank;
	}

	getType(): string {
		return this.type;
	}
}
