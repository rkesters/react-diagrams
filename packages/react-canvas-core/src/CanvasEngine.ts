import { debounce } from 'lodash';
import { CanvasModel } from './entities/canvas/CanvasModel';
import { FactoryBank } from './core/FactoryBank';
import { AbstractReactFactory } from './core/AbstractReactFactory';
import { LayerModel } from './entities/layer/LayerModel';
import { BaseListener, BaseObserver } from './core/BaseObserver';
import { MouseEvent } from 'react';
import { BaseModel } from './core-models/BaseModel';
import { Point } from '@projectstorm/geometry';
import { ActionEventBus } from './core-actions/ActionEventBus';
import { ZoomCanvasAction } from './actions/ZoomCanvasAction';
import { DeleteItemsAction } from './actions/DeleteItemsAction';
import { StateMachine } from './core-state/StateMachine';
export interface CanvasEngineListener extends BaseListener {
	canvasReady(): void;

	repaintCanvas(): void;

	rendered(): void;
}

/**
 * Defines the CanvasEngine options
 */
export interface CanvasEngineOptions {
	registerDefaultDeleteItemsAction?: boolean;
	registerDefaultZoomCanvasAction?: boolean;
	/**
	 * Defines the debounce wait time in milliseconds if > 0
	 */
	repaintDebounceMs?: number;
}

export class CanvasEngine<
	L extends CanvasEngineListener = CanvasEngineListener,
	M extends CanvasModel = CanvasModel
> extends BaseObserver<L> {
	protected model: M ;
	protected layerFactories: FactoryBank<AbstractReactFactory<LayerModel>>;
	protected canvas: HTMLDivElement | undefined;
	protected eventBus: ActionEventBus;
	protected stateMachine: StateMachine;
	protected options: CanvasEngineOptions;

	constructor(model: M, options: CanvasEngineOptions = {}) {
		super();
		this.model = model;
		this.eventBus = new ActionEventBus(this);
		this.stateMachine = new StateMachine(this);
		this.layerFactories = new FactoryBank();
		this.registerFactoryBank(this.layerFactories);

		/**
		 * Overrides the standard options with the possible given options
		 */
		this.options = {
			registerDefaultDeleteItemsAction: true,
			registerDefaultZoomCanvasAction: true,
			repaintDebounceMs: 0,
			...options
		};
		if (this.options.registerDefaultZoomCanvasAction === true) {
			this.eventBus.registerAction(new ZoomCanvasAction({}, this));
		}
		if (this.options.registerDefaultDeleteItemsAction === true) {
			this.eventBus.registerAction(new DeleteItemsAction({}, this));
		}
	}

	getStateMachine() {
		return this.stateMachine;
	}

	getRelativeMousePoint(event: { clientX: number; clientY: number }): Point {
		const point = this.getRelativePoint(event.clientX, event.clientY);
		return new Point(
			(point.x - (this.model?.getOffsetX() ?? 0)) / ((this.model?.getZoomLevel() ?? 1) / 100.0),
			(point.y - (this.model?.getOffsetY() ?? 0)) / ((this.model?.getZoomLevel() ?? 1) / 100.0)
		);
	}

	getRelativePoint(x:number, y: number): Point {
		const canvasRect = this.canvas?.getBoundingClientRect();
		if (!canvasRect) {
			throw new Error(`Can't get relative point because canvas has not been defined.`)
		}
		return new Point(x - canvasRect.left, y - canvasRect.top);
	}

	registerFactoryBank(factory: FactoryBank<any>) {
		factory.registerListener({
			factoryAdded: (event) => {
				event.factory.setDiagramEngine(this);
			},
			factoryRemoved: (event) => {
				event.factory.setDiagramEngine(null);
			}, eventWillFire: ():void => {} , eventDidFire: ():void => {}
		});
	}

	getActionEventBus() {
		return this.eventBus;
	}

	getLayerFactories() {
		return this.layerFactories;
	}

	getFactoryForLayer<F extends AbstractReactFactory<LayerModel>>(layer: LayerModel | string | undefined) {
		if (!layer) { return;}
		if (typeof layer === 'string') {
			return this.layerFactories.getFactory(layer);
		}
		return this.layerFactories.getFactory(layer.getType());
	}

	setModel(model: M) {
		this.model = model;
		if (this.canvas) {
			requestAnimationFrame(() => {
				this.repaintCanvas();
			});
		}
	}

	getModel(): M {
		return this.model;
	}

	repaintCanvas(promise: true): Promise<any>;
	repaintCanvas(): void;
	repaintCanvas(promise?: true): Promise<any> | void {
		const repaintDebounceMs  = this.options.repaintDebounceMs ?? 0;

		/**
		 * The actual repaint function
		 */
		const repaint = () => {
			this.iterateListeners((listener) => {
				if (listener.repaintCanvas) {
					listener.repaintCanvas();
				}
			});
		};

		// if the `repaintDebounceMs` option is > 0, then apply the debounce
		let repaintFn = repaint;

		if (repaintDebounceMs > 0) {
			repaintFn = debounce(repaint, repaintDebounceMs);
		}

		if (promise) {
			return new Promise<void>((resolve) => {
				const l = this.registerListener({
					rendered: () => {
						resolve();
						l.deregister();
					}
				} as L);
				repaintFn();
			});
		}

		repaintFn();
	}

	setCanvas(canvas?: HTMLDivElement | null) {
		if (this.canvas !== canvas) {
			this.canvas = canvas ?? undefined;
			if (canvas) {
				this.fireEvent(undefined, 'canvasReady');
			}
		}
	}

	getCanvas() {
		return this.canvas;
	}

	getMouseElement(event: MouseEvent): BaseModel<any> | null | undefined  {
		return null;
	}

	zoomToFit() {
		if (!this.canvas) {
			throw new Error(`Can not zoom to fit because canvas is not defined`)
		}
		const xFactor = this.canvas.clientWidth / this.canvas.scrollWidth;
		const yFactor = this.canvas.clientHeight / this.canvas.scrollHeight;
		const zoomFactor = xFactor < yFactor ? xFactor : yFactor;

		this.model?.setZoomLevel((this.model?.getZoomLevel() ?? 1) * zoomFactor);
		this.model?.setOffset(0, 0);
		this.repaintCanvas();
	}
}
