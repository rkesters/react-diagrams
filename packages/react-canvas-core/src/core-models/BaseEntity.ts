import * as _ from 'lodash';
import { CanvasEngine } from '../CanvasEngine';
import { BaseEvent, BaseListener, BaseObserver, ParameterFirst } from '../core/BaseObserver';
import { Toolkit } from '../Toolkit';
import { BaseModel } from './BaseModel';

export interface BaseEntityEvent<T extends BaseEntity = BaseEntity> extends BaseEvent {
	entity?: T;
}

export interface LockChangedEvent<T extends BaseEntity = BaseEntity> extends BaseEntityEvent<T>  {
	locked: boolean;
}
export interface BaseEntityListener extends BaseListener {
	lockChanged(event: LockChangedEvent ): void;
}

export type BaseEntityType = 'node' | 'link' | 'port' | 'point';

export interface BaseEntityOptions {
	id?: string;
	locked?: boolean;
}

export type BaseEntityGenerics = {
	LISTENER: BaseEntityListener;
	OPTIONS: BaseEntityOptions;
};

export interface DeserializeEvent<S = ReturnType<BaseEntity['serialize']>> {
	engine: CanvasEngine;
	data: S;
	registerModel(model: BaseModel): void;
	getModel<T extends BaseModel>(id: string | null): Promise<T | null>;
}

type ff<T extends BaseEntityGenerics = BaseEntityGenerics> = ParameterFirst<T['LISTENER']['lockChanged'], BaseEvent>
const FF: ff = {locked: false, entity: {} as BaseEntity} ;
export class BaseEntity<T extends BaseEntityGenerics = BaseEntityGenerics> extends BaseObserver<T['LISTENER']> {
	protected options: T['OPTIONS'];

	protected id: string;
	constructor(options: T['OPTIONS'] = {}) {
		super();
		this.id = options.id ?? Toolkit.UID();
		this.options = {
			id: this.id,
			...options
		};
	}

	getOptions(): T['OPTIONS'] {
		return this.options;
	}

	getID(): string {
		return this.id;
	}

	doClone(lookupTable: { [s: string]: any } = {}, clone: any): void {
		/*noop*/
	}

	clone(lookupTable: { [s: string]: any } = {}) {
		// try and use an existing clone first
		if (lookupTable[this.id]) {
			return lookupTable[this.id];
		}
		let clone = _.cloneDeep(this);
		clone.options = {
			...this.options,
			id: Toolkit.UID()
		};
		clone.clearListeners();
		lookupTable[this.id] = clone;

		this.doClone(lookupTable, clone);
		return clone;
	}

	clearListeners() {
		this.listeners = {};
	}

	deserialize(event: DeserializeEvent<ReturnType<BaseEntity['serialize']>>) {
		this.options.id = event.data.id;
		this.options.locked = event.data.locked;
	}

	serialize() {
		return {
			id: this.options.id,
			locked: this.options.locked
		};
	}

	// fireEvent<Event extends keyof Methods<L, B>,>( k: keyof T['LISTENER'], event: L) {
	// 	super.fireEvent(
	// 		{
	// 			entity: this,
	// 			...event
	// 		},
	// 		k
	// 	);
	// }

	public isLocked(): boolean {
		return this.options.locked ?? false;
	}

	public setLocked(locked: boolean = true) {
		this.options.locked = locked;

		this.fireEvent(
			{ locked: true },
			'lockChanged'
		);
	}
}
