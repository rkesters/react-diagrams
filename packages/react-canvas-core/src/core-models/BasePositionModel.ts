import { BaseModel, BaseModelGenerics, BaseModelListener, BaseModelOptions } from './BaseModel';
import { BaseEntityEvent, DeserializeEvent } from './BaseEntity';
import { Point, Rectangle } from '@projectstorm/geometry';
import { ModelGeometryInterface } from '../core/ModelGeometryInterface';

export interface BasePositionModelListener extends BaseModelListener {
	positionChanged(event: BaseEntityEvent<BasePositionModel>): void;
}

export interface BasePositionModelOptions extends BaseModelOptions {
	position?: Point;
}

export interface BasePositionModelGenerics extends BaseModelGenerics {
	LISTENER: BasePositionModelListener;
	OPTIONS: BasePositionModelOptions;
}

export class BasePositionModel<G extends BasePositionModelGenerics = BasePositionModelGenerics>
	extends BaseModel<G>
	implements ModelGeometryInterface
{
	protected position: Point;

	constructor(options: G['OPTIONS']) {
		super(options);
		this.position = options.position || new Point(0, 0);
	}

	setPosition(point: Point): void;
	setPosition(x: number, y: number): void;
	setPosition(x: Point | number , y?: number | undefined ) {
		this.position = Point.isa(x) ? x : new Point(x,y as number);
		this.fireEvent({}, 'positionChanged');
	}

	getBoundingBox(): Rectangle {
		return new Rectangle(this.position, 0, 0);
	}

	deserialize(event: DeserializeEvent<ReturnType<BasePositionModel['serialize']>>) {
		super.deserialize(event);
		this.position = new Point(event.data.x, event.data.y);
	}

	serialize() {
		return {
			...super.serialize(),
			x: this.position.x,
			y: this.position.y
		};
	}

	getPosition(): Point {
		return this.position;
	}

	getX(): number {
		return this.position.x;
	}

	getY() {
		return this.position.y;
	}
}
