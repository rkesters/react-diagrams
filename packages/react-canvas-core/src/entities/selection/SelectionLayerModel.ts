import { LayerModel } from '../layer/LayerModel';
import { FactoryBank } from '../../core/FactoryBank';
import { AbstractModelFactory } from '../../core/AbstractModelFactory';
import { BaseModel } from '../../core-models/BaseModel';
import { IClientRect } from '../../states/SelectionBoxState';

export class SelectionLayerModel extends LayerModel {
	public box: IClientRect | undefined;

	constructor() {
		super({
			transformed: false,
			isSvg: false,
			type: 'selection'
		});
	}

	setBox(rect: IClientRect) {
		this.box = rect;
	}

	getChildModelFactoryBank(): FactoryBank<AbstractModelFactory<BaseModel>> | null {
		// is not used as it doesnt serialize
		return null;
	}
}
