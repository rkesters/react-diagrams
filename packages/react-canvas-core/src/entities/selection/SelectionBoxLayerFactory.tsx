import * as React from 'react';
import { AbstractReactFactory, GenerateWidgetEvent } from '../../core/AbstractReactFactory';
import { SelectionLayerModel } from './SelectionLayerModel';
import { GenerateModelEvent } from '../../core/AbstractModelFactory';
import { SelectionBoxWidget } from './SelectionBoxWidget';
import { CanvasEngine } from '../..';

export class SelectionBoxLayerFactory extends AbstractReactFactory<SelectionLayerModel> {
	constructor(engine: CanvasEngine) {
		super('selection', engine);
	}

	generateModel(event: GenerateModelEvent): SelectionLayerModel {
		return new SelectionLayerModel();
	}

	generateReactWidget(event: GenerateWidgetEvent<SelectionLayerModel>): JSX.Element {
		return <SelectionBoxWidget rect={event.model.box} />;
	}
}
