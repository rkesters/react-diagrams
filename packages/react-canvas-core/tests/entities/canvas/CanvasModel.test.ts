import { threadId } from 'worker_threads';
import { CanvasEngine, CanvasEngineListener } from '../../../src/CanvasEngine';
import { BaseModel, BaseModelGenerics } from '../../../src/core-models/BaseModel';
import { AbstractModelFactory } from '../../../src/core/AbstractModelFactory';
import { FactoryBank, FactoryBankListener } from '../../../src/core/FactoryBank';
import { CanvasModel, CanvasModelGenerics } from '../../../src/entities/canvas/CanvasModel';
import { LayerModel } from '../../../src/entities/layer/LayerModel';
import {SelectionBoxLayerFactory} from '../../../src/entities/selection/SelectionBoxLayerFactory';

class TestLayerModel extends LayerModel {
	bank = new FactoryBank<
		AbstractModelFactory<
			BaseModel<BaseModelGenerics>,
			CanvasEngine<CanvasEngineListener, CanvasModel<CanvasModelGenerics>>
		>,
		FactoryBankListener<
			AbstractModelFactory<
				BaseModel<BaseModelGenerics>,
				CanvasEngine<CanvasEngineListener, CanvasModel<CanvasModelGenerics>>
			>
		>
	>();
	mockgetChildModelFactoryBank = jest.fn();
	getChildModelFactoryBank(
		engine: CanvasEngine<CanvasEngineListener, CanvasModel<CanvasModelGenerics>>
	): FactoryBank<
		AbstractModelFactory<
			BaseModel<BaseModelGenerics>,
			CanvasEngine<CanvasEngineListener, CanvasModel<CanvasModelGenerics>>
		>,
		FactoryBankListener<
			AbstractModelFactory<
				BaseModel<BaseModelGenerics>,
				CanvasEngine<CanvasEngineListener, CanvasModel<CanvasModelGenerics>>
			>
		>
	> {
		this.mockgetChildModelFactoryBank(engine);
		return this.bank;
	}

	m = [new BaseModel({}), new BaseModel({ selected: true })];
	getSelectionEntities = jest.fn().mockImplementation(() => this.m);
	getModels = jest.fn().mockImplementation(() => this.m);
}

describe('CanvasModel', () => {
	test.each([
		{ name: 'Default', options: undefined },
		{
			name: 'With options',
			options: {
				offsetX: 100,
				offsetY: 200,
				zoom: 43,
				gridSize: 10
			}
		}
	])('constructor $name', (tc) => {
		const model = new CanvasModel(tc.options);

		expect(model.getOptions()).toMatchSnapshot({
			id: expect.any(String)
		});
	});

	test.each([{ name: 'Default', options: undefined }])('layes $name', (tc) => {
		const model = new CanvasModel(tc.options);
		const layer = new TestLayerModel();

		model.addLayer(layer);
		expect(model.getLayers().length).toEqual(1);
		expect(model.getSelectionEntities()).toEqual(layer.m);
		expect(layer.getSelectionEntities).toBeCalled();

		expect(model.getSelectedEntities()).toEqual([layer.m[1]]);

		model.clearSelection();
		expect(model.getSelectedEntities()).toEqual([]);

		expect(model.getModels()).toEqual(layer.m);

		const layer2 = new TestLayerModel();
		model.addLayer(layer2);

		model.removeLayer(layer);
		expect(model.getLayers().length).toEqual(1);
		expect(model.getLayers()).toEqual([layer2]);

		model.setLocked();
		expect(model.isLocked()).toBeTruthy();
		model.setLocked(false);
		expect(model.isLocked()).toBeFalsy();
		model.setLocked(true);
		expect(model.isLocked()).toBeTruthy();

		model.setOffset(1000, 2000);
		expect(model.getOffsetX()).toEqual(1000);
		expect(model.getOffsetY()).toEqual(2000);

		model.setOffsetX(-100);
		expect(model.getOffsetX()).toEqual(-100);
		expect(model.getOffsetY()).toEqual(2000);
		model.setOffsetY(-200);
		expect(model.getOffsetX()).toEqual(-100);
		expect(model.getOffsetY()).toEqual(-200);
		model.setOffset(0, 0);

		model.setZoomLevel(200);
		expect(model.getZoomLevel()).toEqual(200);

		model.setGridSize(0);
		expect(model.getOptions().gridSize).toEqual(0);

		expect(model.getGridPosition(1)).toEqual(1);
		model.setGridSize(100);
		expect(model.getGridPosition(1)).toEqual(0);
		expect(model.getGridPosition(101)).toEqual(100);
	});

	test.each([{ name: 'Default', options: undefined }])('serialize $name', (tc) => {
		const model = new CanvasModel(tc.options);
		const layer = new TestLayerModel({type: 'selection' });

		model.addLayer(layer);

		const serial = model.serialize();

		expect(serial).toMatchSnapshot({ id: expect.any(String), layers: [{ id: expect.any(String) }] });

		serial.gridSize = 100;

		const engine = new CanvasEngine();
		engine.getLayerFactories().registerFactory(new SelectionBoxLayerFactory());

		model.deserialize( {data: serial, engine} as any);

		expect(model.getOptions().gridSize).toEqual(100);

		serial.gridSize = 300;
		model.deserializeModel(serial, engine);
		expect(model.getOptions().gridSize).toEqual(300);

	});
});
