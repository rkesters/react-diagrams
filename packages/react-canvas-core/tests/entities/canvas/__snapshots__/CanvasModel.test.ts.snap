// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CanvasModel constructor Default 1`] = `
Object {
  "gridSize": 0,
  "id": Any<String>,
  "offsetX": 0,
  "offsetY": 0,
  "zoom": 100,
}
`;

exports[`CanvasModel constructor With options 1`] = `
Object {
  "gridSize": 10,
  "id": Any<String>,
  "offsetX": 100,
  "offsetY": 200,
  "zoom": 43,
}
`;

exports[`CanvasModel serialize Default 1`] = `
Object {
  "gridSize": 0,
  "id": Any<String>,
  "layers": Array [
    Object {
      "extras": undefined,
      "id": Any<String>,
      "isSvg": undefined,
      "locked": undefined,
      "models": Object {},
      "selected": undefined,
      "transformed": undefined,
      "type": "selection",
    },
  ],
  "locked": undefined,
  "offsetX": 0,
  "offsetY": 0,
  "zoom": 100,
}
`;
